### PR TITLE
[Merged by Bors] - feat: port LinearAlgebra.Vandermonde

### PR DESCRIPTION
**Pull request by EmilieUthaiwat**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Vandermonde.lean [hunk 1]
```diff
@@ -106,70 +114,71 @@ theorem det_vandermonde {n : ℕ} (v : Fin n → R) :
...
+  simp only [det_vandermonde_eq_zero_iff, Ne.def, not_exists, not_and, Classical.not_not]
 #align matrix.det_vandermonde_ne_zero_iff Matrix.det_vandermonde_ne_zero_iff
 
 theorem eq_zero_of_forall_index_sum_pow_mul_eq_zero {R : Type _} [CommRing R] [IsDomain R] {n : ℕ}
     {f v : Fin n → R} (hf : Function.Injective f)
     (hfv : ∀ j, (∑ i : Fin n, f j ^ (i : ℕ) * v i) = 0) : v = 0 :=
   eq_zero_of_mulVec_eq_zero (det_vandermonde_ne_zero_iff.mpr hf) (funext hfv)
-#align matrix.eq_zero_of_forall_index_sum_pow_mul_eq_zero Matrix.eq_zero_of_forall_index_sum_pow_mul_eq_zero
+#align matrix.eq_zero_of_forall_index_sum_pow_mul_eq_zero
+Matrix.eq_zero_of_forall_index_sum_pow_mul_eq_zero
```
Comment (semorrison) [inline Mathlib/LinearAlgebra/Vandermonde.lean:166]: I remember there's some script which expects #align statements to be all on one line, and there's an exception in the long line linter for exactly this. Could you put these back on one line?

File: Mathlib/LinearAlgebra/Vandermonde.lean [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  unfold vandermonde
+  induction' n with n ih
+  · exact det_eq_one_of_card_eq_zero (Fintype.card_fin 0)
+  calc
+    det (of fun i j : Fin n.succ => v i ^ (j : ℕ)) =
+        det
+          (of fun i j : Fin n.succ =>
+            Matrix.vecCons (v 0 ^ (j : ℕ)) (fun i => v (Fin.succ i) ^ (j : ℕ) - v 0 ^ (j : ℕ)) i) :=
+      by
+      refine' det_eq_of_forall_row_eq_smul_add_const (Matrix.vecCons 0 1) 0 (Fin.cons_zero _ _) _
```
Comment (semorrison) [inline Mathlib/LinearAlgebra/Vandermonde.lean:89]: At very least we should have a porting note here noting that this is much more complicated than the mathlib3 proof.

Possibly we could factor it out as a lemma? This calc block is getting out of hand.

Comment (semorrison) [reply Mathlib/LinearAlgebra/Vandermonde.lean:89]: Oh, I see! In the mathlib3 proof, the calc block left two goals for later (the curly brace blocks after the `calc`). You can do this in mathlib4 as well, but you need to use `?_` instead for the goals you want to postpone.

File: Mathlib/LinearAlgebra/Vandermonde.lean [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+      intro i j
+      simp_rw [of_apply]
+      rw [Matrix.cons_val_zero]
+      refine' Fin.cases _ (fun i => _) i
+      · simp
+      rw [Matrix.cons_val_succ, Matrix.cons_val_succ, Pi.one_apply]
+      ring
```
Comment (semorrison) [inline Mathlib/LinearAlgebra/Vandermonde.lean:96]:
```suggestion
      det_eq_of_forall_row_eq_smul_add_const (Matrix.vecCons 0 1) 0 (Fin.cons_zero _ _) ?_
```

File: Mathlib/LinearAlgebra/Vandermonde.lean [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        apply det_eq_of_forall_col_eq_smul_add_pred fun _ => v 0
+        · intro j
+          simp
+        · intro i j
+          simp only [smul_eq_mul, Pi.add_apply, Fin.val_succ, Fin.coe_castSucc, Pi.smul_apply]
+          rw [Finset.sum_range_succ, add_comm, tsub_self, pow_zero, mul_one, Finset.mul_sum]
+          congr 1
+          refine' Finset.sum_congr rfl fun i' hi' => _
+          rw [mul_left_comm (v 0), Nat.succ_sub, pow_succ]
+          exact Nat.lt_succ_iff.mp (Finset.mem_range.mp hi')
```
Comment (semorrison) [inline Mathlib/LinearAlgebra/Vandermonde.lean:136]:
```suggestion
    _ = (Finset.prod Finset.univ (fun i => v (Fin.succ i) - v 0)) *
    det fun i j : Fin n => v (Fin.succ i) ^ (j : ℕ) := congr_arg _ ?_
```

File: Mathlib/LinearAlgebra/Vandermonde.lean (hunk 2 above)
Comment (EmilieUthaiwat) [reply Mathlib/LinearAlgebra/Vandermonde.lean:89]: I did not know I could postpone a goal in mathlib4 as well. Thank you!

**Pull request closed without merging**